### PR TITLE
Add Stripe Payment Request Button support

### PR DESCRIPTION
**Pull request by jranks123**
## Why are you doing this?
Now that the Stripe contract has been updated to take Google Pay into account, we (almost) can allow the Payment Request Button to be used to take one off payments. Woohoo!

| The Stripe Payment Request Button in all it's glory |
|------------|
|stripe|

### CODE REVIEW

File: assets/pages/new-contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx [hunk 1]
```diff
@@ -132,20 +131,29 @@ function onClick(event, props: PropTypes) {
...
-
 // The value of result will either be:
 // . null - browser has no compatible payment method)
 // . {applePay: true} - applePay is available
-// . {applePay: false} - GooglePay or PaymentRequestApi available
-const browserIsCompatible = (result: Object) => {
-  if (weAreSupportingGooglePay) {
-    return result !== null;
+// . {applePay: false} - GooglePay, Microsoft Pay and PaymentRequestApi available
+const availablePaymentRequestButtonPaymentMethod: Object => (StripePaymentMethod | null) = (result: Object) => {
```
Comment (joelochlann) [inline assets/pages/new-contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: don't we normally use the other return type notation?

```javascript
const availablePaymentRequestButtonPaymentMethod = (result: Object): StripePaymentMethod | null =>
...
```

I find this one hard to parse since I'm used to the other one - and it seems you need to notate the `Object` input type twice which I'm not a fan of

Comment (jranks123) [reply assets/pages/new-contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: 👍 sorry about this, not sure what came over me...

File: assets/pages/new-contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx [hunk 2]
```diff
@@ -156,18 +164,13 @@ function initialisePaymentRequest(props: PropTypes) {
...
 
   paymentRequest.canMakePayment().then((result) => {
-    if (browserIsCompatible(result)) {
-      trackComponentClick('apple-pay-loaded');
-      props.setCanMakeStripePaymentRequestPayment(true);
+    const paymentMethod = availablePaymentRequestButtonPaymentMethod(result);
+    if (paymentMethod !== null) {
+      trackComponentClick(`${paymentMethod}-loaded`);
+      props.setPaymentRequestButtonPaymentMethod(paymentMethod);
+      setUpPaymentListener(props, paymentRequest, paymentMethod);
```
Comment (joelochlann) [inline assets/pages/new-contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: I just wonder if to eliminate race conditions it's worth doing `setUpPaymentListener(props, paymentRequest, paymentMethod);` before `props.setPaymentRequestButtonPaymentMethod(paymentMethod);` since I think the latter might be entirely synchronous (i.e. the state update & DOM re-render all happen before the next line)

it's super-paranoid but I'm just wondering if there's a way someone might manage to click the button before the event handler's bound. could be best just to make that impossible

Comment (joelochlann) [reply assets/pages/new-contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: > since I think the latter might be entirely synchronous (i.e. the state update & DOM re-render all happen before the next line)

actually I'm really not sure about that...

Comment (joelochlann) [reply assets/pages/new-contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: I think it might be at the React level that the state changes are decoupled from re-renders: https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous

Comment (jranks123) [reply assets/pages/new-contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: Yeah sure, why not! No harm done
